File: Projects/leetcode/WinnerSquareGame.java
package leetcode;

class WinnerSquareGame {
    public static boolean winnerSquareGame(final int n) {
        final boolean[] winners = new boolean[n + 1];
        for (int i = 0; i <= n; i++) {
            if (winners[i]) {
                continue;
            }
            for (int k = 1; i + k * k <= n; k++) {
                winners[i + k * k] = true;
            }
        }
        return winners[n];
    }
}

File: Projects/leetcode/winnerSquareGame.kt
package leetcode

internal object WinnerSquareGame {
    fun winnerSquareGame(n: Int): Boolean {
        val winners = BooleanArray(n + 1)
        for (i in 0..n) {
            if (winners[i]) {
                continue
            }
            var k = 1
            while (i + k * k <= n) {
                winners[i + k * k] = true
                k++
            }
        }
        return winners[n]
    }
}